### ISSUE TITLE
Missing result locations for builtin types

### ISSUE THREAD

**Issue by vesim987** [ziglang/zig#16211]
### Zig Version

0.11.0-dev.3826+7a197f124

### Steps to Reproduce and Observed Behavior

While porting microzig to latest zig version I have found couple of missing result locations for builtins.

```zig
const Foo = struct {
    bar: u32,
};

const MyStruct = struct {
    foo: *Foo,
};

export fn issue_1() void {
    _ = MyStruct{
        .foo = &.{
            .bar = @intCast(20),
        },
    };
}

fn do_something(_: *Foo) void {}

export fn issue_2() void {
    do_something(&.{
        .bar = @intCast(20),
    });
}


export fn isse_3 () void {
  _ = [1]u16{
      @intCast(20),
  };
}
```
It also applies to other builtins.

### Expected Behavior

Every example to compile.

**Comment by nektro** [comment]
`issue_1` and `issue_2` create a `*const Foo` in their bodies, so i would not expect those to compile

**Comment by mlugg** [comment]
It makes sense that `issue_1` and `issue_2` don't work, but perhaps we want them to. `issue_3` should certainly work however - #16226 is technically a duplicate but we can track that issue there for separation.

**Comment by perillo** [comment]
@mlugg, Issue #16226 has been closed.

I suspect that the problem is no longer "missing result locations", but about "how to cast away constness.".

For issue 2, however, I think that the solution is simply:

```zig
fn do_something(_: *const Foo) void {}
```

**Comment by mlugg** [comment]
Yes, #16226 has been closed because the `issue_3` snippet now works. The other issues detailed here were resolved by #16512, so this issue can now be closed. The constness cast failure you noticed is just a minor bug in the given test case, unrelated to the issue at hand.

**Comment by perillo** [comment]
> The constness cast failure you noticed is just a minor bug in the given test case, unrelated to the issue at hand.

What I assumed from your comment:
> It makes sense that issue_1 and issue_2 don't work, but perhaps we want them to

is that since the *destination* is not a const pointer, then the compiler should automatically cast away the constness, since it is safe.
